overlord/snapstate/handlers.go

```diff
@@ -3467,25 +3530,57 @@ func InjectAutoConnect(mainTask *state.Task, snapsup *SnapSetup) {
 	mainTask.Logf("added auto-connect task")
 }
 
-// GetSnapDirOptions returns *dirs.SnapDirOptions configured according to the
-// enabled experimental features. The state must be locked by the caller.
-var GetSnapDirOptions = getSnapDirOptions
+type HiddenDirOptions struct {
+	// UseHidden states whether the user has requested that the hidden data dir be used
+	UseHidden bool
+
+	// MigratedToHidden states whether the data has been migrated to the hidden dir
+	MigratedToHidden bool
+}
 
-func getSnapDirOptions(state *state.State) (*dirs.SnapDirOptions, error) {
-	tr := config.NewTransaction(state)
+// GetSnapDirOpts return the options required to get the correct snap
+// data dir, based on the state's options.
+func (o *HiddenDirOptions) GetSnapDirOpts() *dirs.SnapDirOptions {
+	return &dirs.SnapDirOptions{
+		HiddenSnapDataDir: o.MigratedToHidden,
+	}
+}
 
+// GetHiddenDirOpts checks if the feature flag is set and if the snap data
+// has been migrated, first checking the SnapSetup (if not nil) and then
+// the SnapState. The state must be locked by the caller.
+var GetHiddenDirOpts = func(st *state.State, snapsup *SnapSetup, name string) (*HiddenDirOptions, error) {
```

Comment:
are there places outside of snapstate that needs or will need the full HiddenDirOptions?

otherwise my preference would still be to have two helpers:

* `func getHiddenDirOpts(st *state.State, snapsup *SnapSetup/* ? */, name string) (*HiddenDirOptions, error)` internal to snapstate
* `func GetSnapDirOpts(st *state.State, snapsup *SnapSetup/* ? */, name string) (*SnapDirOptions, error)` exported

the first would need mocking support only in export_test.go

Reply:
Yes, `snapshotstate` also uses the HiddenDirOptions. However, I can do `var getHiddenDirOpts = snapstate.GetHiddenDirOpts` in snapshotstate and then have two mocking functions, one in snapshotstate's export_test.go and another in snapstate's. The hiddenDirOpts getter is still exported but at least there's no exposed mocking function in snapstate (although it is duplicated).

Reply:
maybe I'm confused I see it only using it to get SnapDirOptions atm? or is that code in future PRs?

Reply:
Ah wait, I misunderstood. I can decouple the two and non-snapstate things don't need to get the hidden opts (maybe migration opts would be a better name) and go directly to the dir opts. That makes more sense, I'll do that. Thanks